[Feat] 모듈 구조 변경 및 토큰 로직

choijungp opened the pull request:
## 🌁 Background
<!-- 해당 PR을 작성하게 된 이유를 적어주세요. -->

07월 25일 긴 .. 회의 끝에 결정된 Network, Persistence 모듈 부시기를 진행하였습니다.  
그 외에도 토큰 관련 로직, 리팩토링 필요한 로직들을 정리하였습니다.    



## 👩‍💻 Contents
<!-- 작업 내용을 적어주세요 -->
- NetworkService, Persistence 모듈 제거 후 DataSource 모듈에 병합
- Then 외부 라이브러리 의존성 제거 및 사용중인 코드 수정
- TokenManager 구현 (토큰 저장, 불러오기, 삭제 담당)
- Endpoint에 isAuthorized 값 추가하여 Endpoint header 값에 accessToken 저장
- UserDataRepository에 토큰 재발급 로직 (reissueToken 추가)  
- 로그인 시 소셜 로그인 타입, 프로필 이미지 저장
- 온보딩 서버 Request Enum 값 수정 
- SceneDelegate에서 로그인 여부에 따라 분기 처리



## ✅ Testing
<!-- 테스트 방법을 적어주세요 -->

로그인 후 재접속 했을 시에 로그인 화면이 아닌 탭바 진입을 확인했습니다.  


https://github.com/user-attachments/assets/b30a6186-b553-4eaf-b075-af5136ba631e




## 📝 Review Note
<!-- PR과정에서 든 생각이나 개선할 내용이 있다면 적어주세요. -->

**1. 소셜 로그인 타입 저장**      
와이어프레임과 달리 확정된 마이페이지에서 소셜 로그인 타입을 보여주고 있지 않습니다.      
하지만 혹시 모를 확장을 위해 소셜 타입 저장을 구현하였습니다. (빼도 되겠다고 판단되면 빼겠습니다.)   

<br>

**2. 카카오 SDK 유저 정보 가져오기**      
Apple 로그인과 다르게 카카오 로그인에서는 닉네임을 빼오지 않아 따로 저장하는 로직이 없었습니다.   
그래서 마이 페이지에서 유저 정보를 보여줄 때 카카오 로그인 시 유저 정보를 빼오지 못하는 이슈가 있었습니다.  

AuthRepository에 fetchKakaoUserInfo 함수를 통해 닉네임과 프로필 이미지 URL을 빼와 저장하는 로직을 구현하였습니다.    

프로필 이미지 URL의 경우 URL 타입으로 UserDefaults에 저장하지 못하여서 absoluteString을 저장합니다.   
추후 빼서 사용할 때 참고해주세용  !!!

그리고 ... 서버에서 유저 정보를 보내주는 API를 파준다고 합니다 !!!! (현재는 닉네임만 이야기 되었는데 프로필 이미지도 얘기해봐야 할 듯)    
그럼 아마 UserDefaults에서 빼오는 현재와 달리 로직이 달라질 수 있을 것 같아요 .. 


<br>

**⭐️⭐️⭐️ 3. Endpoint의 isAuthorized값에 따른 헤더 값 처리 ⭐️⭐️⭐️**

Endpoint에 isAuthorized 값을 추가하여 해당 값이 true라면 헤더에 Authorization 값을 넣어주는 로직을 추가하였습니다.   
하지만 이때 TokenManager에서 accessToken을 빼오는 로직에서 throw가 가능하지만 이것은 계산 프로퍼티로 .. try 로직을 어떻게 처리할지 모르겠어서 일단 강제 언래핑을 하도록 하였습니다. ..
 
```swift
var headers: [String : String] {
    var headers: [String: String] = [
        "Content-Type": "application/json",
        "accept": "*/*"
    ]

    if isAuthorized {
        let accessToken = try! TokenManager.shared.loadToken(tokenType: .accessToken)
        headers["Authorization"] = "Bearer \(accessToken)"
    }

    switch self {
    case .login(_, _, let token):
        headers["SocialAccessToken"] = token
    case .reissue(let refreshToken):
        headers["Refresh-Token"] = refreshToken
    default:
        break
    }

    return headers
}
```


혹시 좋은 방법이 있을까요 ??
Repository에서 endpoint 생성 후 isAuthorized 값에 따라 endpoint의 header 값에 엑세스 토큰을 추가할까도 고민했지만,

like this .... 
```swift
func logout() async throws {
    let endpoint = AuthEndpoint.logout
    if endpoint.isAuthorized {
        let accessToken = try TokenManager.shared.loadToken(tokenType: .accessToken)
        endpoint.headers["Authorization"] = "Bearer \(accessToken)"
    }
    _ = try await networkService.request(endpoint: endpoint, type: String.self)
    try tokenManager.removeToken()
}
```

그럼 서버 통신하는 Endpoint에서 동일한 로직 (= isAuthorized 확인 후 헤더 값 설정)이 중복되는 것 같아 일단 강제 언래핑으로 진행하였습니다 .. 

<br>

**⭐️⭐️⭐️ 4. 401 토큰 재발급 + retry ⭐️⭐️⭐️**

NetworkService에서 401 받으면 토큰 재발급하고, 재발급한 토큰으로 retry하는 로직을 아직 넣지 않았어요 ...
띵 델리게이트에게 부탁드립니다. .. 요로시쿠 ! 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


## Summary by CodeRabbit

* **신규 기능**
  * 토큰 관리 및 사용자 정보 저장을 위한 싱글톤 기반의 TokenManager, UserDefaultsStorage 클래스가 추가되었습니다.
  * 토큰 재발급 기능이 UserDataRepository에 추가되어 자동 로그인 흐름이 개선되었습니다.
  * 인증 관련 에러 및 사용자 정보 저장/불러오기 에러에 대한 상세 에러 메시지가 추가되었습니다.
  * SplashView가 도입되어 앱 실행 시 초기 스플래시 화면이 표시됩니다.

* **기능 개선**
  * 카카오/애플 로그인 시 사용자 닉네임, 프로필 이미지 등 추가 정보가 저장됩니다.
  * 토큰 전달 방식이 개선되어, 네트워크 요청 시 내부적으로 토큰을 처리합니다.
  * 의존성 주입 구조가 간소화되고, 일부 모듈(NetworkService, Persistence 등)이 제거되었습니다.
  * 온보딩 및 인증 엔드포인트에서 토큰 파라미터가 제거되어 간결해졌습니다.
  * 온보딩 완료 후 홈 화면 진입 방식이 변경되어 직접 루트 뷰 컨트롤러를 설정합니다.

* **버그 수정**
  * 토큰 및 사용자 정보 저장/삭제 로직의 신뢰성과 에러 처리가 강화되었습니다.

* **스타일/리팩터**
  * 외부 라이브러리(Then) 사용이 제거되고, UI 속성 설정 방식이 직접 할당 방식으로 변경되었습니다.
  * 여러 UI 컴포넌트 설정 코드에서 `do` 블록이 제거되어 가독성이 향상되었습니다.

* **문서화**
  * 프로토콜 및 일부 코드의 주석이 최신 흐름에 맞게 수정되었습니다.

* **기타**
  * 불필요한 의존성, 모듈, 프로토콜 및 관련 파일들이 프로젝트에서 제거되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

choijungp committed:
Refactor: NetworkService, Persistence 모듈 DataSource 모듈 합치기

choijungp committed:
Refactor: Then 외부 라이브러리 제거 및 사용 코드 제거

choijungp committed:
Feat: TokenManager 구현

choijungp committed:
Refactor: Endpoint에 isAuthorized 값 추가

- Endpoint에 isAuthorized 값 추가 (인증이 필요한지 여부)
- 그렇다면 Endpoint 만들 때 header에 TokenManager에서 accessToken 빼와서 저장

choijungp committed:
Refactor: UserDataRepository에 token 재발급 로직 추가

choijungp committed:
Feat: 로그인 시 소셜 로그인 타입, 프로필 이미지 저장

- 카카오 로그인 시 nickname, profileImageUrl을 KakaoSDK에서 받아와 저장하도록 로직 추가
- 카카오, 애플 로그인 시 소셜 로그인 타입 저장하도록 로직 추가

choijungp committed:
Fix: 온보딩 서버 Request Enum 값 변경

- 온보딩 서버 Request Enum 값 변경
- 온보딩 UI 수정 (SubTitle 여부)

choijungp committed:
Feat: UserDataRepository에서 토큰 재발급 가능 여부에 따라 앱 진입점 분기처리

choijungp committed:
Fix: 토큰 재발급 로직 Response 타입 변경

- TokenResponseDTO 구현
- reissue response 타입 교체

choijungp committed:
Style: 불필요한 주석 제거

choijungp committed:
Style: UserDataRepository 오타 수정

taipaise commented on Projects/App/Sources/SceneDelegate.swift:
```diff
@@ -19,13 +20,21 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-        window.makeKeyAndVisible()
-
-        self.window = window
+        Task {
```
네트워크 통신이 오래 걸린다면, 해당 작업이 진행되는 동안 아마 흰색이나 검은색 화면이 보이는 경우가 있을 것 같습니다!
추후 lauch screen과 같은 역할을 하는 splash 화면을 구현해두고, token reissue가 끝나면 탭바/인트로 뷰 중 하나를 보여주는 건 어떨까요?

choijungp replied:
오 !!! 그럴 수 있겠군요 !!!  그럼 이렇게 수정하는게 맞을까요 ?!?!?!? 🥺

```swift
func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
    guard let windowScene = (scene as? UIWindowScene) else { return }
    let window = UIWindow(windowScene: windowScene)

    DIContainer.shared.dependencyInjection()

    guard let userDataRepository = DIContainer.shared.resolve(type: UserDataRepositoryProtocol.self)
    else { fatalError("userDataRepository 의존성이 등록되지 않았습니다.") }

    // 요기가 추가된 splash 화면을 보여주는 로직입니다 !!!!
    window.rootViewController = SplashView()
    window.makeKeyAndVisible()
    self.window = window

    Task { @MainActor in
        let isLogined = await userDataRepository.reissueToken()
        if isLogined {
            window.rootViewController = TabBarView()
        } else {
            let introView = IntroView()
            let navigationController = UINavigationController(rootViewController: introView)
            window.rootViewController = navigationController
        }
    }
}
```

taipaise replied:
넵 요론식으로 가면 좋을거 같습니다!👍👍

taipaise commented on Projects/DataSource/Sources/Common/DataSourceDependencyAssembler.swift:
깔끔깔끔합니다~!~!

taipaise commented on Projects/DataSource/Sources/Endpoint/AuthEndpoint.swift:
```diff
@@ -30,37 +30,38 @@ extension AuthEndpoint: Endpoint {
             "accept": "*/*"
         ]
 
+        if isAuthorized {
+            let accessToken = try! TokenManager.shared.loadToken(tokenType: .accessToken)
+            headers["Authorization"] = "Bearer \(accessToken)"
+        }
```
와 PR에서 해당 문제에 대해 언급한 것도 파악해서 알려주네요.. 래빗님 대단하네요

choijungp replied:
오 싯 !!!!! 래빗 똑똑하군요 !!!!! 

그럼 NetworkService에서 endpoint 받고, isAuthorized에 대해 헤더값 설정하게 할까요 ?? 
like this !!!?
```swift
func request<T: Decodable>(endpoint: Endpoint, type: T.Type) async throws -> T? {
    var request = try endpoint.makeURLRequest()
    if endpoint.isAuthorized {
        let accessToken = try TokenManager.shared.loadToken(tokenType: .accessToken)
        request.headers["Authorization"] = "Bearer \(accessToken)"
    }

    let (data, response) = try await URLSession.shared.data(for: request)
    // 네트워크 로직 ..
}
```

taipaise commented on Projects/DataSource/Sources/Endpoint/OnboardingEndpoint.swift:
```diff
@@ -8,8 +8,8 @@
 enum OnboardingEndpoint {
-    case registerOnboarding(accessToken: String, choices: [String: String])
-    case registerRecommendedRoutine(accessToken: String, selectedRoutines: [Int])
+    case registerOnboarding(choices: [String: String])
```
온보딩 등록 요청 시, 키 값으로 어떤 값들이 들어갈까요?.? 만약에 만약에 enum 과 같은 타입으로 범위를 정해둘 수 있다면, 타입을 강제하는 방법은 어떨까요??

choijungp replied:
엇 그 OnboardingUseCase에서 enum으로 받은 값들을 dictionary로 바꿔주는 로직을 수행하고 있습니다 !!

```swift
// OnboardingUseCase.swift
public func registerOnboarding(onboardingChoices: [OnboardingChoiceType]) async throws -> [RecommendedRoutineEntity] {
    let choices = convertToDictionary(onboardingChoices: onboardingChoices)
    let recommendedRoutines = try await onboardingRepository.registerOnboarding(onboardingChoices: choices)
    return recommendedRoutines
}

private func convertToDictionary(onboardingChoices: [OnboardingChoiceType]) -> [String: String] {
    var result: [String: String] = [:]
    let onboardingTypes: [OnboardingType] = [.time, .frequency, .feeling, .outdoor]
    for type in onboardingTypes {
        guard let choice = onboardingChoices.first(where: { $0.onboardingType == type })
        else { break }
        result[choice.onboardingType.key] = choice.value
    }
    return result
}
```

호옥시 이 부분을 말씀하신게 맞을까요 ?? 그렇다면 UseCase가 아닌 Repository에서 변경하는게 맞다고 생각하시나욤 ... ??

choijungp replied:
근데 또 이용약관 로직보면 ... 

UseCase에서는 다음과 같이 repository 호출하고, DataSource 단에서 바꿔주고 있긴 하네요 .... 
```swift
// LoginUseCase.swift
public func sumbitAgreement(agreements: [TermsType: Bool]) async throws {
    try await authRepository.submitAgreement(agreements: agreements)
}


// DataSource .... LoginEndpoint.swift
var bodyParameters: [String : Any] {
    switch self {
    case .login(let socialLoginType, let nickname, _):
        // ···
    case .agreements(let agreements):
        var parameters: [String: Any] = [:]
        for agreement in agreements {
            parameters[agreement.key.termKey] = agreement.value
        }
        return parameters
    default:
        return [:]
    }
}

extension TermsType {
    var termKey: String {
        switch self {
        case .service: "agreedToTermsOfService"
        case .privacy: "agreedToPrivacyPolicy"
        case .age: "isOverFourteen"
        }
    }
}
```

근데 그렇다 보니 TermsType의 위치가 애매하긴 해서 LoginEndpoint 파일 밑에 그냥 넣어둔 ㄹㅈㄷ 멋대로 짠 코드였습니다 ......


해당 로직 어디서 바꿔줄지 결정하면 하나로 통일하면서 수정하겠습니다 !!

taipaise replied:
말씀해주신 상황은 진짜 개발을 진행하면 많이 발생하는 문제인거 같아요.. 저도 하다보면 애매한 겅우가 많아서 일단 진행하는 경우가 많아서 왕 공감합니다 🥹

말씀해주신 내용이 맞아요!! 괜찮으시다면 변환은 repository에서 하는건 어떤가요? 가능하면 데이터 변환 로직은 reposiorty에서 진행하면 좋을 것 같다고 생각합니다. domain에 있는 usecase들은 데이터를 요청할 때나/데이터를 받을 때나 해당 데이터가 어떻게 가공해야할지 상관 쓰지 않고, 핵심 비즈니스 로직을 실행하기 위해 필요한 데이터 타입을 repository에게 주거나/받거나 할 수 있으면 좋을것 같다 생각하기 때문입니다!! 조이 의견은 어떠신가요?

choijungp replied:
하아 ......... 그것은 너무 고민되는 사항 ....... 
띵의 의견 뭔말인지 너무 이해가고 ... 동의하지만 그럼 UseCase에서 어떤 것을 할 수 있을까요 ......

넘 passthrough 되는 기분이 들기도 합니다 ㅠ !!!!!!!!!!!!!

taipaise replied:
참고할 수 있을만한 다른 레포나 다른 분들의 조언도 정말 절실하네요 🥹 
이후 시니어 개발자 분의 피드백도 한 번 듣고 고민해볼까요.?.?

choijungp replied:
오키 조아요 !!!!!!! 그럼 일단 현상 유지 하고 돌아오는 토요일에 조언을 구해볼까요 ??

그 외의 코드 리뷰 사항은 [37bef52](https://github.com/YAPP-Github/Bitnagil-iOS/pull/24/commits/37bef529baf17c45c1a95d00fd0c70391031ee1b)에서 반영했습니다 !!!

taipaise replied:
넵 그래주시면 압도적 감사입니다!!!!

taipaise commented on Projects/DataSource/Sources/Persistence/KeychainStorage.swift:
```diff
@@ -1,23 +1,19 @@
+//  DataSource
 //
-//  Created by 반성준 on 6/21/25.
+//  Created by 최정인 on 7/26/25.
```
옳게 된 주석 인정합니다 👍

choijungp replied:
ㅋㅋㅋ 아 고의아니게 ㅋㅋㅋㅋㅋ 코드만 복붙해서 이렇게 됐네요 ㅎㅎㅋㅋ ... 😅

taipaise commented on Projects/DataSource/Sources/Repository/AuthRepository.swift:
```diff
@@ -111,6 +93,28 @@ final class AuthRepository: AuthRepositoryProtocol {
+                }
+            }
+
+            Task { @MainActor in
```
카카오의 UserApi를 MainActor 실행 환경에서 동작하도록 해주신것 같은데요, 이유가 궁금합니다!!
카카오 문서에서는 관련 내용을 찾지 못했는데, 메인 스레드에서 동작해야하는 제약조건이 있는걸까요?

choijungp replied:
오오  .... 사실 `fetchKakaoToken()` 함수를 복붙하고 수정하느라 그대로 메인 스레드에서 실행하도록 했습니다 ..... 😢   
매우매우 죄송합니다 .............. 

흠 ................ 사실 굳이 메인 스레드에서 동작하지 않아도 큰 문제는 없을 것 같은 녀석이긴 합니다 !! 
그래서 또 요 부분에서 보면 유저 정보를 빼오는 로직이지만 error를 던지는 것은 kakaoTokenFetchFailed를 던져주고 있습니다 ...... (매우 죄송합니다 22)   

다음과 같이 수정하려고 하는데 어떠신가욤 !!!!!!!!

```swift
private func fetchKakaoUserInfo() async throws -> (nickname: String, profileImageUrl: URL) {
    try await withCheckedThrowingContinuation { continuation in
        let resultHandler: (User?, Error?) -> Void = { user, error in
            if let error {
                continuation.resume(throwing: AuthError.unknown(error))
            } else if
                let nickname = user?.kakaoAccount?.profile?.nickname,
                let profileImageUrl = user?.kakaoAccount?.profile?.profileImageUrl {
                continuation.resume(returning: (nickname, profileImageUrl))
            } else {
                continuation.resume(throwing: AuthError.kakaoUserInformationFetchFailed)
            }
        }

        UserApi.shared.me(completion: resultHandler)
    }
}
```

taipaise replied:
개발자의 기본 소양 중 하나가 복붙 아닐까요 ㅌㅋㅋㅋㅌ 저도 많이 하는 실수라,,,
수정된 코드는 너무 좋습니다~!

taipaise reviewed (commented):
1. 소셜 로그인 타입 저장 -> 좋은 생각입니다~!
2. 서버에서 유저 정보를 주면 오히려 좋을거 같아요!! 참고하겠습니다.
3. 전혀 생각지도 못한 부분이네요.!.! 다음과 같은 방법은 어떤지 조심스럽게 말씀드려봅니다,, 
    - head가 딕셔너리라 key:value 형식으로 데이터 저장,.
    - 토큰 load 오류 발생 시 -> header에 "error"라는 키로 토큰을 가지고 오지 못했다는 값을 넣어주기 (ex: "error" : can’t load token 등)
    - networkService 객체에서 요청 전송 전 header에 error를 키로 하는 value가 있는지 확인. 있다면 네트워크 요청 보내지 않고 error throw
    - repositiory에서 networkService에서 뱉은 에러를 받아 철
4. 요 부분은 차주 안에 꼭 로직 추가해두겠읍니다

구조 변경이 많아 많이 귀찮으셨을텐데 고생하셨습니다!!

choijungp committed:
Refactor: 코드 리뷰 반영

- NetworkService에서 endpoint.isAuthorized 값 확인하여 헤더에 토큰 값 세팅해주는 로직으로 수정
- 카카오 유저 정보 (닉네임, 프로필 이미지 URL) 메인스레드에서 동작하는 것 제거
- SceneDelegate에서 SplashView를 먼저 표시하고, reissueToken 결과에 따라 rootViewController 교체하도록 구조 개선

choijungp committed:
Refactor: import SnapKit 추가

choijungp commented on Projects/Presentation/Sources/Common/SplashView.swift:
```diff
@@ -0,0 +1,32 @@
+//  Created by 최정인 on 7/27/25.
+//
+
+import UIKit
+
```
https://github.com/YAPP-Github/Bitnagil-iOS/commit/e00a90d40c77d7873d5d6ed37203bae0d2eaabdc 에서 수정했습니다 .. .근데 왤케  review in progress 시간이 길어요 ?? 네 ??

taipaise reviewed (approved):
오늘도 정말 고생 많으셨습니다~!~!
